### PR TITLE
eth, internal/web3ext: implement storageRangeAt

### PR DESCRIPTION
**Pull request by obscuren**
StorageAt is currently limited and requires needless iterations due to
the limitation of the trie iterator. At present we can't start
iterating from any given key, thus we need to loop over any key
that's not inclusive in the range of start and end.

### CODE REVIEW

**Comment by obscuren** [comment]
```javascript
var h = eth.getBlock(eth.blockNumber-2).hash;
var a = "0xFA7B9770Ca4cb04296Cac84F37736d4041251CDF";
var s = "0x0000000000000000000000000000000000000000000000000000000000000001";
var e = "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";
debug.storageAt(h, 0, a, s, e, 100)
```

returns:

```
{
  0x0000000000000000000000000000000000000000000000000000000000000001: "0x0000000000000000000000000000000000000000000000000000000000000000",
  0x0000000000000000000000000000000000000000000000000000000000000008: "0x0000000000000000000000000000000000000000000000000000000000000000",
  0x1632fdd11d2f42e69bbb9114d9cdb236c6e3dde438ce903a816e54a5c534f730: "0x0000000000000000000000000000000000000000000000000000000000000000",
...
  0xf3f7a9fe364faab93b216da50a3214154f22a0a2b415b23a84c8169e8b636f00: "0x0000000000000000000000000000000000000000000000000000000000000000",
  0xf3f7a9fe364faab93b216da50a3214154f22a0a2b415b23a84c8169e8b636f01: "0x0000000000000000000000000000000000000000000000000000000000000000"
}
```

**Comment by fjl** [comment]
Please rename to storageRangeAt

**Comment by zelig** [comment]
@obscuren this should work in the light client odr as well right?

File: eth/api.go [hunk 1]
```diff
@@ -547,3 +547,87 @@ func (api *PrivateDebugAPI) TraceTransaction(ctx context.Context, txHash common.
...
+		// Mutate the state if we haven't reached the tracing transaction yet
+		if idx < txIndex {
+			vmenv := vm.NewEnvironment(context, stateDb, api.config, vm.Config{})
+			_, _, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(tx.Gas()))
+			if err != nil {
+				return nil, fmt.Errorf("mutation failed: %v", err)
+			}
+			stateDb.DeleteSuicides()
+			continue
+		}
```
Comment (fjl) [inline eth/api.go]: Please extract this code into its own function. This basically recomputes the state up to a certain transaction
and is duplicated in the tracing code just above.

File: eth/api.go [hunk 2]
```diff
@@ -547,3 +547,87 @@ func (api *PrivateDebugAPI) TraceTransaction(ctx context.Context, txHash common.
...
+
+		stateObject := stateDb.GetStateObject(contractAddress)
+		// We need to check if the object exists again. It might have been deleted in between the
+		// transactions.
+		if stateObject != nil {
+			trie := stateObject.GetTrie(api.eth.ChainDb())
+			it := trie.Iterator()
+
+			for it.Next() && len(result.Storage) < maxResult {
```
Comment (fjl) [inline eth/api.go]: IMHO it would be better to use ForEachStorage here instead of exposing the trie.

**Comment by fjl** [comment]
I've done the updates myself. The current version is pretty useless because it keeps the first n keys that we're hit by the random iteration order. It'd be better to keep a sorted buffer so the function always returns the n lowest keys instead. I'll add that next.

**Comment by karalabe** [comment]
Postponed to 1.5.6

**Comment by fjl** [comment]
@obscuren Please review this.

The API has changed and is now correct in cases where pre-images of storage keys are missing. In your initial implementation, such keys would override the entry for 0x0000..00. I have discussed this change with @chriseth and @yann300.

The other bug that I fixed is the one mentioned above: your implementation kept the first N matching keys. This caused random 'gaps' in the output, making the API pretty useless for actually syncing storage to the client. The new implementation returns the first N keys from the `start key`. Clients can use this to download full storage incrementally by requesting from 0x00 with limit 100, then using the last key + 1 as the start key for the next call.

Example (on ropsten):

```text
POST http://127.0.0.1:8545/

{
   "jsonrpc": "2.0",  "id": 1, "method": "debug_storageRangeAt",
   "params":[
       "0xced82545131cdb6253f183a62dc3eaf81259be5cc744767798bcfb95664dc32c",
       4, 
       "0xB4A5BB81C6962Fa87Af6886ABE59b3cc56ec44b8",
       "0x", 
       "0x50",
       8
    ]
}
```

```text

**Comment by fjl** [comment]
Now that I think about it, the API still isn't right. It works reasonably well when key pre-images are available. When they're not, all keys will be 0x0000..00 (but hashedKey will be different). Since the key range parameters operate on `key` instead of `hashedKey`, the client can't reliably skip over keys it has already downloaded. IMHO it would be better to change the API so the key range applies to `hashedKey`. We should probably also make the `end` parameter exclusive instead of inclusive.

@chriseth @yann300 please comment. This API is for remix so the result should make sense for whatever remix is doing.

**Comment by chriseth** [comment]
If the actual keys are not available, the storage dump is only of very limited use to the debugger. If you are only in some "exploration" view where you know where the variables are in storage, it still makes sense, but you will always have the feeling of wandering in the dark.

Ranges in hashed values make no sense, because the parts you are interested in will not be contiguous. I think if the actual keys are not available, your only chance is to download everything (in chunks) and check on the client ("frontend").

Futhermore, the problem is that we never know which areas in storage are used. Take for example an array of size 1000000. It does not cost anything to allocate that. If we have the keys, then we can request that range easily. Most values will be zero because storing items is expensive, so the query result will still be short. If we do not have the keys, though, our only hope is to request the full storage, compute every single hash value in the frontend and check whether the storage dump contains that value...

So in the long run, I would really prefer if the database had all sha3 preimages. Can we acutally literally use "getNodeData" for that, or does it something else than inverting sha3 over the network?

Concerning this PR: I think if the actual key is not available, we should use `null` for the `key` property. Furthermore, items with unknown keys should sort just as `0` sorts, but before any element that has a known key. If the key is not available for any item in the contract storage (not only the query response), we should flag that (meaning that the result might be missing some elements for a range query) using e.g. `someKeysUnknown: true'.

**Comment by fjl** [comment]
We have discussed this offline, with the following result:

The API will change yet again. `debug_storageRangeAt` will take parameters `blockHash`, `txIndex`, `account`, `startKey`, `limit`. The `startKey` will apply to the hashed key instead of the preimage. Remix will use this to sync full storage to the browser. Returned keys will still contain the preimage if available.

An additional call, `debug_preimage`, will allow retrieving the preimage of any trie key. To make storage debugging actually useful, the call will also be extended to retrieve preimages of the SHA3 EVM instruction. A command line flag to enable instrumentation of the VM in this way will be added.
The instrumentation will be on by default in `--dev` mode.

**Comment by chriseth** [comment]
Concerning `debug_preimage`: Perhaps we could rephrase it as "... can be used to retrieve the preimage of a given SHA3 image, if available. Conforming implementations SHOULD at least provide preimages of any trie node and any SHA3 operation computed by the EVM, unless some syncing method was used that does not provide that information'.

**Comment by chriseth** [comment]
@gumb0 could you please adjust the implementation in cpp-ethereum to the new specification?

**Comment by karalabe** [comment]
@fjl Is still still slated for 1.5.6 or do we push it out to 1.5.7? I'd like to finish the release merges today.

**Comment by yann300** [comment]
@fjl instead of returning an array for the storage,
is that possible to return a map of hashed key?
```
"storage": {
 "0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563":
      {
        "key": "0x0000000000000000000000000000000000000000000000000000000000000000",
        "hashedKey": "0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563",
        "value": "0x0000000000000000000000945e237bf1540eea146f5e830593d1d54c279616a9"
      },
...
}
```

**Comment by fjl** [comment]
Yes

**Comment by yann300** [comment]
@fjl is the return value is sorted by hashed keys?

**Comment by fjl** [comment]
This is now #14350

**Pull request closed without merging**